# HG changeset patch
# User Daniel J Sebald <daniel.sebald@ieee.org>
# Date 1464883616 18000
#      Thu Jun 02 11:06:56 2016 -0500
# Node ID a4c744194bbb825e543059953e980072c7893616
# Parent  1955af16b42e16ff138cb86fd934276494db0c80
Add equal data aspect ratio to gnuplot graphics toolkit 3D plots (bug #47974).

* axis.m: Clarify meaning of 'square' and 'equal' in documentation.  Remove
  gnuplot-specific FIXME.

* __gnuplot_draw_axes__.m: Remove line of cruft for sz_str.  If the plot is 3D
  and data aspect ratio is equal in at least x-y dimensions, then set view
  equal.

diff --git a/scripts/plot/appearance/axis.m b/scripts/plot/appearance/axis.m
--- a/scripts/plot/appearance/axis.m
+++ b/scripts/plot/appearance/axis.m
@@ -56,20 +56,20 @@
 ## @noindent
 ## turns tic marks on for all axes and tic mark labels on for the y-axis only.
 ##
 ## @noindent
 ## The following options control the aspect ratio of the axes.
 ##
 ## @table @asis
 ## @item @qcode{"square"}
-## Force a square aspect ratio.
+## Force a square axis aspect ratio.
 ##
 ## @item @qcode{"equal"}
-## Force x distance to equal y-distance.
+## Force x-axis unit distance to equal y-axis (and z-axis) unit distance.
 ##
 ## @item @qcode{"normal"}
 ## Restore default aspect ratio.
 ## @end table
 ##
 ## @noindent
 ## The following options control the way axis limits are interpreted.
 ##
@@ -186,24 +186,16 @@ function limits = __axis__ (ca, ax, vara
     elseif (strcmpi (ax, "image"))
       __axis__ (ca, "equal");
       set (ca, "plotboxaspectratiomode", "auto");
       __do_tight_option__ (ca);
     elseif (strcmpi (ax, "square"))
       set (ca, "dataaspectratiomode", "auto",
                "plotboxaspectratio", [1, 1, 1]);
     elseif (strcmp (ax, "equal"))
-      if (strcmp (get (ancestor (ca, "figure"), "__graphics_toolkit__"), "gnuplot"))
-        ## FIXME: gnuplot applies the aspect ratio activepostionproperty.
-        set (ca, "activepositionproperty", "position");
-        ## The following line is a trick used to trigger the recalculation of
-        ## aspect related magnitudes even if the aspect ratio is the same
-        ## (useful with the x11 gnuplot terminal after a window resize)
-        set (ca, "dataaspectratiomode", "auto");
-      endif
       set (ca, "dataaspectratio", [1, 1, 1], "plotboxaspectratio", [5 4 4]);
 
     elseif (strcmpi (ax, "normal"))
       ## Set plotboxaspectratio to something obtuse so that switching
       ## back to "auto" will force a re-calculation.
       set (ca, "plotboxaspectratio", [3 2 1]);
       set (ca, "plotboxaspectratiomode", "auto",
                "dataaspectratiomode", "auto");
diff --git a/scripts/plot/util/private/__gnuplot_draw_axes__.m b/scripts/plot/util/private/__gnuplot_draw_axes__.m
--- a/scripts/plot/util/private/__gnuplot_draw_axes__.m
+++ b/scripts/plot/util/private/__gnuplot_draw_axes__.m
@@ -121,28 +121,40 @@ function __gnuplot_draw_axes__ (h, plot_
       sz_str = sprintf ("%s %.15g, %.15g;\n", sz_str, pos(3), pos(4));
     endif
   else ## activepositionproperty == outerposition
     fprintf (plot_stream, "unset tmargin;\n");
     fprintf (plot_stream, "unset bmargin;\n");
     fprintf (plot_stream, "unset lmargin;\n");
     fprintf (plot_stream, "unset rmargin;\n");
     fprintf (plot_stream, "set origin %g, %g;\n", pos(1:2));
-    sz_str = "";
     if (strcmpi (axis_obj.dataaspectratiomode, "manual"))
       sz_str = sprintf ("ratio %g", -dr);
     else
       sz_str = "noratio";
     endif
     sz_str = sprintf ("set size %s %g, %g;\n", sz_str, pos(3:4));
   endif
   if (! isempty (sz_str))
     fputs (plot_stream, sz_str);
   endif
 
+  ## Code above uses axis size for the data aspect ratio, which isn't
+  ## quite correct.  The only fine control is to set all axes units equal.
+  if (nd == 3 &&
+      strcmpi (axis_obj.dataaspectratiomode, "manual") &&
+      axis_obj.dataaspectratio(1) == axis_obj.dataaspectratio(2))
+    if (axis_obj.dataaspectratio(1) == axis_obj.dataaspectratio(3))
+      zstr = "z";
+    else
+      zstr = "";
+    endif
+    fprintf (plot_stream, "set view equal xy%s;\n", zstr);
+  endif
+
   ## Reset all labels, axis-labels, tick-labels, and title
   ## FIXME: We should have an function to initialize the axis.
   ##        Presently, this is dispersed in this function.
   fputs (plot_stream, "unset label;\n");
   fputs (plot_stream, "unset xtics;\n");
   fputs (plot_stream, "unset ytics;\n");
   fputs (plot_stream, "unset ztics;\n");
   fputs (plot_stream, "unset x2tics;\n");
